app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts

```diff
@@ -66,82 +72,74 @@ export const mapReducer = (state: MapState, action: BaseAction): MapState => {
       for (const row of addDataToMapAction.payload.datasets.data.rows) {
         //Retrieving the feature in the row
         const addedFeature: MapFeature = row[0].properties;
-        //Looping through all features stored in the redux state currently
-        for (const resultFeature of result.features) {
-          //if the uri of the addedFeature and a feature in the redux state match
-          if (resultFeature.uri === addedFeature.uri) {
-            //Change the state of the feature to RENDERED, because we are putting it on the map
-            resultFeature.state = MapFeatureState.RENDERED;
-            //Change the dirty variable for that feature type to false, as it is obvious we are now adding
-            //features of that type to the map.
-            result.featuresByType[
-              resultFeature.type! as keyof typeof result.featuresByType
-            ].dirty = false;
-            console.debug(
-              "changed map feature " +
-                resultFeature.uri +
-                " to state " +
-                MapFeatureState.RENDERED
-            );
-          }
+        //if the uri of the addedFeature and a feature in the redux state match
+        const resultFeature = result.features.find(
+          ({uri}) => uri === addedFeature.uri
+        );
+        if (!resultFeature) {
+          throw Error(
+            "Features are not added to the map until after they are loaded,\
+              and therefore it should exist in the features list on the state.\
+              (Look at code ADD_MAP_FEATURES reducer case)"
+          );
         }
+
+        //Change the state of the feature to RENDERED, because we are putting it on the map
+        resultFeature.state = MapFeatureState.RENDERED;
+        //Change the dirty variable for that feature type to false, as it is obvious we are now adding
+        //features of that type to the map.
+        result.featuresByType[
+          resultFeature.type! as keyof typeof result.featuresByType
+        ].dirty = false;
+        console.debug(
+          "changed map feature " +
+            resultFeature.uri +
+            " to state " +
+            MapFeatureState.RENDERED
+        );
+
         /*
         Check the filterState of the type of feature being added to the map.
         Loop throrugh the attributes of the feature and check to see which are of type number.
```

Comment:
If you've got multiple pieces of logic here, can you separate them into functions to make it more obvious? (And call them from the action handler.)